initramfs-init.in

```diff
@@ -279,6 +279,23 @@ is_url() {
 	esac
 }
 
+# Do some tasks to make sure mounting the ZFS pool is A-OK
+prepare_zfs_root() {
+	local _root_vol=${KOPT_root#ZFS=}
+	local _root_pool=${_rool_vol%%/*}
+
+	# Force import if this has been imported on a different system previously
+	if [ "$KOPT_zfs_force" = 1 ]; then
+		zpool import -f $_root_vol
+	fi
+
+	# Ask for encryption password
+	if [ $(zpool list -H -o feature@encryption $_root_pool) = "active" ]; then
```

Comment:
this check will return true if you have any encrypted dataset on this pool, while root dataset might be unencrypted.
in this case `_encryption_root` will be set to `-` and load-key will fail.
It's better to specifically check for encryption on this dataset
`"$(zfs get -H -o value encryption "$_root_vol")" != "off"`

Reply:
Ah, thanks for the pointer. I guess that even if the root fs isn't encrypted we should unlock the pool though (e.g. if /home is encrypted the user most likely does want us to decrypt it). I'll fix the _encryption_root setting though.

Reply:
1. There can be multiple encryption roots
2. Looks like initramfs doesn't mount any dataset except root, which means user will have to mount encrypted datasets manually (zfs mount -a for example)

I'd say leave decryption of non root datasets to openrc

Reply:
OpenRC can mount those with zfs mount -a but it's not interactive so it can't ask for the password.